### PR TITLE
[WIP] Add bounding box feature

### PR DESCRIPTION
**Pull request by jstpierref**
# PR details
This PR refers to issue #331. It adds a `bounding_box` output feature to the library. This way the user can train models with `bounding_box` output like it is encountered in object detection. 

Currently, the only implemented measure is IOU (Intersection-Over-Union). The current default loss is the smoothed L1 loss (also called Huber loss) used by the Fast R-CNN algorithm. I plan to add the YOLO v3 localization loss as well. The following snippet runs flawlessly, but I might have forgotten key additions to make this feature usable so feel free to change it/let me know!

Basic test that I ran:
```python
from ludwig.api import LudwigModel
from sklearn import datasets
import pandas as pd 
import numpy as np

iris = datasets.load_iris()
training_dataframe = pd.DataFrame(iris['data'], columns=['a','b','c','d'])
training_dataframe['target'] = list(map(lambda x: np.array([x,x,x,x]), iris['target']))

# train a model
model_definition = {
	'input_features': [
		{'name': 'a', 'type': 'numerical'},
		{'name': 'b', 'type': 'numerical'},
		{'name': 'c', 'type': 'numerical'},
		{'name': 'd', 'type': 'numerical'}
	], 
	'output_features': [{'name': 'target', 'type': 'bounding_box'}]
}
model = LudwigModel(model_definition)
train_stats = model.train(training_dataframe)
predictions = model.predict(training_dataframe)
model.close()
```

# Documentation 
Doc not updated yet

### CODE REVIEW

**Comment by w4nderlust** [comment]
@nitrofortin did a first pass, you're definitely on the right track, thank you for contributing this!
Will do a second pass and leave some specific comments.
One generic thing: usually before committing we run pycharm code refactoring for optimizing imports and whitespaces etc, could you please do the same (in pycharm's ui when you commit there's a box in the top right of the widnow that does all the clode cleanup and checks).

**Comment by nitrofortin** [comment]
Yes sure!

File: ludwig/features/boundingbox_feature.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    @staticmethod
+    def add_feature_data(
+            feature,
+            dataset_df,
+            data,
+            metadata,
+            preprocessing_parameters,
+    ):
+        data[feature['name']] = np.stack(dataset_df[feature['name']],
```
Comment (w4nderlust) [inline ludwig/features/boundingbox_feature.py]: How do you expect the content of the column containing the bounding box to look like? I imageine something like this:
```
image_path,bbox
path.jpeg,10 20 15 15
```
so the bbox column should contain 4 values divided by space.
In this case, I believe the processing should be different from what you are doing, but maybe you are thinking about another format.

File: ludwig/features/boundingbox_feature.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+                self.name
+            )
+
+            iou_val = get_bbox_iou(
+                targets,
+                predictions,
+                self.name
+            )
+
+            mean_average_precision = None
```
Comment (w4nderlust) [inline ludwig/features/boundingbox_feature.py]: usually in ludwig we don't compute means in the graph because when we evaluate we want the means to be computed on the whole set rather than on the specific batch. There is a mechanism for collecting all the outputs of the batches when evaluating and then performa the average just one time. look at the code in model.py that uses OutputConfig for more details.

File: ludwig/features/boundingbox_feature.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+    output_config = OrderedDict([
+        (LOSS, {
+            'output': EVAL_LOSS,
+            'aggregation': SUM,
+            'value': 0,
+            'type': MEASURE
+        }),
+        ('mean_iou', {
+            'output': 'iou',
```
Comment (w4nderlust) [inline ludwig/features/boundingbox_feature.py]: 'iou' should be a constant

File: ludwig/features/boundingbox_feature.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        if PREDICTIONS in result and len(result[PREDICTIONS]) > 0:
+            postprocessed[PREDICTIONS] = result[PREDICTIONS]
+            if not skip_save_unprocessed_output:
+                np.save(
+                    npy_filename.format(name, PREDICTIONS),
+                    result[PREDICTIONS]
+                )
+            del result[PREDICTIONS]
+
+        if PROBABILITIES in result and len(result[PROBABILITIES]) > 0:
```
Comment (w4nderlust) [inline ludwig/features/boundingbox_feature.py]: this can probably be removed

File: ludwig/models/modules/measure_modules.py [hunk 5]
```diff
@@ -158,3 +158,25 @@ def r2(targets, predictions, output_feature_name):
     r2 = tf.subtract(1., res_ss / tot_ss,
                      name='r2_{}'.format(output_feature_name))
     return r2
+
+
+def bbox_iou(targets, predictions, output_feature_name):
+    x11, y11, x12, y12 = tf.split(targets, 4, axis=1)
```
Comment (w4nderlust) [inline ludwig/models/modules/measure_modules.py]: Not sure why are you transposing few lines from here, can you add a comment on what shapes do you expcet the inputs to have?

File: ludwig/features/boundingbox_feature.py (hunk 1 above)
Comment (nitrofortin) [reply ludwig/features/boundingbox_feature.py]: It is currently accepting bounding boxes as np arrays, but I agree that 4 values divided by space  is more convenient especially when dealing with files. I guess it means the bbox values would be handled as strings in python, right?

File: ludwig/features/boundingbox_feature.py (hunk 4 above)
Comment (nitrofortin) [reply ludwig/features/boundingbox_feature.py]: Done

File: ludwig/features/boundingbox_feature.py (hunk 3 above)
Comment (nitrofortin) [reply ludwig/features/boundingbox_feature.py]: Done

File: ludwig/models/modules/measure_modules.py (hunk 5 above)
Comment (nitrofortin) [reply ludwig/models/modules/measure_modules.py]: I removed the transposed variables.

**Comment by nitrofortin** [comment]
Pushed some changes following your comments! Let me know if there are other requirements you think this PR should match.

File: ludwig/models/modules/measure_modules.py (hunk 5 above)
Comment (w4nderlust) [reply ludwig/models/modules/measure_modules.py]: Was looking into literature, it seems like it is more common to provide bounding boxes as: x, y h, w where x and y are the coordinates of the center, h w are height and width respectively.
Can you please doublecheck a couple implementations like faster rcnn and yolo to make sure we are doing the most common thing?

Comment (nitrofortin) [reply ludwig/models/modules/measure_modules.py]: Yes, I will have look over the week

**Comment by w4nderlust** [comment]
> Pushed some changes following your comments! Let me know if there are other requirements you think this PR should match.

Looks good, thank you!
I would doublecheck the format of the bounding boxes (x y h w versus x1 y1 x1 y2) in the most common datasets for object detection. You could also potentially add it as a parameter of the decoder and implement both functions, but that we can do later.
I guess now a good test would be to test this on a simple dataset to confirm it works, and add that test to the test suite.
Finally documentation needs to be added in the user guide.
Thank you so much again for the work your are putting into this!

**Comment by w4nderlust** [comment]
@nitrofortin any update on this?

**Comment by nitrofortin** [comment]
@w4nderlust Still need to test it on a real example. And you are right, (x y h w) seems more common. To make it work we will need a feature extractor, and I am still looking into how Ludwig does this automatically. If you have any insights, let me know!

**Comment by nitrofortin** [comment]
@w4nderlust Still need to test it on a real example. And you are right, (x y h w) seems more common. To make it work we will need a feature extractor, and I am still looking into how Ludwig does this automatically. If you have any insights, let me know!

**Comment by ifokeev** [comment]
@nitrofortin up, please

**Comment by carlogrisetti** [comment]
I could have some time to spare in trying to pull this in... Do you (@nitrofortin \ @w4nderlust) know what's still missing?
I will be doing another boxing network in the near future (I hope it's days from now) and would really like to leverage this WIP PR to find out if this is a viable way to go.

I'll try to tinker with it, but any help is appreciated

File: ludwig/models/modules/measure_modules.py (hunk 5 above)
Comment (carlogrisetti) [reply ludwig/models/modules/measure_modules.py:169]: Moreover, it seems to me that the bounding box output feature has been placed in the input features. Am I reading that correctly?

**Comment by nitrofortin** [comment]
Been a while since I last worked on this. I'll check and let you know.

**Comment by w4nderlust** [comment]
> I could have some time to spare in trying to pull this in... Do you (@nitrofortin \ @w4nderlust) know what's still missing?
> I will be doing another boxing network in the near future (I hope it's days from now) and would really like to leverage this WIP PR to find out if this is a viable way to go.
> 
> I'll try to tinker with it, but any help is appreciated

It would be great if you could help @carlogrisetti , but if you can I would hold on until we complete the TF2 porting, as things may look different then. We are almost there, probably a couple more weeks. Is that a good time for you?

**Comment by carlogrisetti** [comment]
Oh great news regarding the TF2 porting... I can surely see the value in doing this after the fact :)
It might not be aligned with my timetable but I think TF2 porting is more important than a minor rework on my side.

**Comment by w4nderlust** [comment]
If you have time snd willingness can I suggest you to help with the tf2
porting instead? I can point you to specific self contained tasks that
would be easy but require care.

On Fri, 29 May 2020 at 00:16, Carlo Grisetti <notifications@github.com>
wrote:

> Oh great news regarding the TF2 porting... I can surely see the value in
> doing this after the fact :)
> It might not be aligned with my timetable but I think TF2 porting is more
> important than a minor rework on my side.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uber_ludwig_pull_344-23issuecomment-2D635805046&d=DwMCaQ&c=r2dcLCtU9q6n0vrtnDw9vg&r=RrWxefUdVFeb7Dpg6EF4Ug&m=mJiXFn31X7CFFdi5o6KA6xYleI6MiA4gb0V_zEGbCU4&s=dGAyO56M-cJAvmKBuBj9stNQR6ytVRHKXJhVLXl63vY&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AACVISG6YSYRIHEBW6PLTBLRT5OMZANCNFSM4HNFYAYA&d=DwMCaQ&c=r2dcLCtU9q6n0vrtnDw9vg&r=RrWxefUdVFeb7Dpg6EF4Ug&m=mJiXFn31X7CFFdi5o6KA6xYleI6MiA4gb0V_zEGbCU4&s=z7Bj4L8E5LX6k463aKzh3aeD6AF4L5SOA026A-HcYqs&e=>
> .
>
-- 
----
Piero Molino
Sr Research Scientist
UBER AI Labs

**Comment by carlogrisetti** [comment]
Willingness, yes... time, maybe (it really depends upon daily tasks)... skill... well let's see...
The only concern is that it might be way over my level, since I have really limited experience with TF\Keras (hence my love for ludwig), but I can always try. Maybe the more tedious tasks can be the best suited ones...

Throw something at me and we shall see :)

**Comment by w4nderlust** [comment]
> Throw something at me and we shall see :)

Give me an email or other means for reaching out to you and I can give you all the tedious tasks you want :)

**Comment by gustavorps** [comment]
Any updates about this feature?

**Comment by carlogrisetti** [comment]
I focused elsewhere, since I was able to effectively obtain the same result using an output feature in the form of a vector with TRBL pixel positions (but I think any notation would have been effective), since I start from the idea that the object I am looking for is always present and it's only matter of determining where
